cmd/dp-import-reporter/main.go

```diff
@@ -0,0 +1,100 @@
+package main
+
+import (
+	"context"
+	"os"
+	"os/signal"
+	"runtime/debug"
+	"syscall"
+
+	"github.com/ONSdigital/dp-import-reporter/config"
+	"github.com/ONSdigital/go-ns/kafka"
+	"github.com/ONSdigital/go-ns/log"
+	"github.com/coocood/freecache"
+	"github.com/ONSdigital/dp-import-reporter/event"
+	"github.com/ONSdigital/dp-import-reporter/client"
+	"net/http"
+	"io/ioutil"
+	"io"
+	"github.com/ONSdigital/dp-import-reporter/server"
+	"github.com/ONSdigital/dp-import-reporter/message"
+)
+
+func main() {
+	log.Namespace = "dp-event-reporter"
+
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, os.Interrupt, syscall.SIGTERM)
+	errorChannel := make(chan error, 1)
+
+	cfg, err := config.Get()
+	if err != nil {
+		os.Exit(1)
+	}
+
+	log.Info("dp-import-reporter config", log.Data{
+		"config": cfg,
+	})
+
+	datasetAPIClient, err := client.NewDatasetAPIClient(cfg.DatasetAPIURL, cfg.DatasetAPIAuthToken, &http.Client{}, ResponseBodyReader{})
+	if err != nil {
+		os.Exit(1)
+	}
+
+	cache := freecache.NewCache(cfg.CacheSize)
+	// TODO why is this set?
+	debug.SetGCPercent(20)
+
+	server.Start(cache, cfg.BindAddress, errorChannel)
+
+	reportEventHandler := event.Handler{
+		Cache:         cache,
+		DatasetAPI:    datasetAPIClient,
+		ExpireSeconds: cfg.CacheExpiry,
+	}
+
+	eventReceiver := &event.Receiver{
+		Handler: reportEventHandler,
+	}
+
+	// create the report event kafka kafkaConsumer.
+	kafkaConsumer, err := kafka.NewConsumerGroup(cfg.Brokers, cfg.ReportEventTopic, log.Namespace, kafka.OffsetNewest)
+	if err != nil {
+		log.ErrorC("unexpected error while attempting to create kafka kafkaConsumer", err, nil)
+		os.Exit(1)
+	}
+
+	// create event kafkaConsumer.
+	reportEventConsumer := message.NewMessageConsumer(kafkaConsumer, eventReceiver)
+	reportEventConsumer.Listen()
+
+	// shutdown all the things.
+	gracefulShutdown := func() {
+		log.Info("Attempting graceful shutdown of service", nil)
+
+		ctx, cancel := context.WithTimeout(context.Background(), cfg.GracefulShutdownTimeout)
+		defer cancel()
+
+		server.Shutdown(ctx)
```

Comment:
Should the service not close the consumers first then the server as the service will consume more messages whilst the server is being shutdown.

Luckily any failure to send a request to an API will result in the message to not be committed to the consumer group, so technically the ordering here should not be a problem, but if this gets extended in anyway, ordering the closing of connections could become important.

Reply:
good spot, I've moved the consume close to be the first step of the graceful shut down